File: testing/match_scratch.js
// let row = 3;
// let col = 3;
// let text = "11";
// let combi = (text.match(/1/gi) || []).length;
// let key_name = ["5", "10", "50", "100", "500", "1k", "10k", "100k"];
// let key_lists = {};
// for (let i = 0; i < key_name.length; i++) {
//     key_lists[key_name[i]] = 0;
// }
// let table_match = [];
// for (let i = 0; i < 8; i++) {
//     table_match.push(Object.assign({}, key_lists));
// }
// let result_table = [];
// for (let i = 0; i < row; i++) {
//     let row = [];
//     for (let j = 0; j < col; j++) {
//         row.push(null);
//     }
//     result_table.push(row);
// }
// if (combi > 0) {
//     for (let i = 0; i < table_match.length; i++) {
//         for (let j = 0; j < key_name.length; j++) {
//             table_match[i][key_name[j]] = (i === j) ? 3 : 0;
//         }
//     }
//     let select_table = Object.entries(table_match[(combi > table_match.length) ? table_match.length - 1 : combi - 1]);
//     let newsets = [];
//     for (let i = 0; i < select_table.length; i++) {
//         if (select_table[i][1] !== 0) {
//             newsets = Array(select_table[i][1]).fill(select_table[i][0]); 
//         }
//     }
    
//     let prev_dup = [];
//     let count = 0;
//     while (count < newsets.length) {
//         let trows = Math.floor(Math.random() * result_table.length);
//         let tcols = Math.floor(Math.random() * result_table[trows].length);
//         let position = `rows=${trows},cols=${tcols}`;
//         if (!prev_dup.includes(position)) {
//             result_table[trows][tcols] = newsets[count];
//             prev_dup.push(position);
//         } else {
//             count -= 1;
//         }
//         count += 1;
//     }
// }

// console.log(result_table)

let row = 3;
let col = 4;
let text = "1";
let combi = (text.match(/1/gi) || []).length;
let key_name = ["5", "10", "50", "100", "500", "1k", "10k", "100k"];
let key_lists = Object.fromEntries(key_name.map(k => [k, 0]));
let table_match = Array.from({ length: 8 }, () => ({ ...key_lists }));
let result_table = Array.from({ length: row }, () => Array.from({ length: col }, () => null));

if (combi > 0) {
    for (let i = 0; i < table_match.length; i++) {
        for (let j = 0; j < key_name.length; j++) {
            table_match[i][key_name[j]] = i === j ? 3 : 0;
        }
    }

    let select_table = Object.entries(table_match[combi > table_match.length ? table_match.length - 1 : combi - 1]);
    let newsets = [];
    for (let i = 0; i < select_table.length; i++) {
        if (select_table[i][1] !== 0) {
            newsets = Array(select_table[i][1]).fill(select_table[i][0]);
        }
    }

    let prev_dup = [];
    let count = 0;
    while (count < newsets.length) {
        let trows = Math.floor(Math.random() * result_table.length);
        let tcols = Math.floor(Math.random() * result_table[trows].length);
        let position = `rows=${trows},cols=${tcols}`;

        if (!prev_dup.includes(position)) {
            result_table[trows][tcols] = newsets[count];
            prev_dup.push(position);
        } else {
            count--;
        }
        count++;
    }
} 
 
let storeCombi = result_table.flat(); 
for (let i = 0; i < result_table.length; i++) {
  let row = result_table[i]; 
  for (let rt = 0; rt < row.length; rt++) { 
    let ranKeyName = key_name[Math.floor(Math.random() * key_name.length)]; 
    if (row[rt] === null) { 
      if (storeCombi.filter(name => name === ranKeyName).length < 2) { 
        result_table[i][rt] = ranKeyName; 
        storeCombi.push(ranKeyName);
      } else {
        rt--;
      }
    }
  }
}

console.log(result_table);

File: testing/match_scratch.py
import re
from pprint import pprint 
import random
from itertools import chain

row = 3
col = 3
text = "0"
combi = len(re.findall("1", text, flags= re.I))

key_name = ['5', '10', '50', '100', '1k', '2k', '20k', '200k'] 
key_lists = {k: 0 for _, k in enumerate(key_name)}
table_match = [key_lists.copy() for i in range(8)]
result_table = [[ None for _ in range(int(col))] for _ in range(int(row))]
if combi > 0 :
    for i, tm in enumerate(table_match):
        for ik, km in enumerate(key_name):
            table_match[i][km] = i == ik and 3 or 0

    select_table = table_match[combi > len(table_match) and len(table_match) - 1 or combi - 1].items()
    for i, stl in enumerate(select_table):
        if stl[1] !=0:
            select_table = [stl[0] for _ in range(stl[1])]
    prev_dup = []
    count = 0
    while count < len(select_table):
        trows = random.randrange(0, len(result_table))
        tcols = random.randrange(0, len(result_table[trows]))
        position = f"rows={trows},cols={tcols}" 
        if position not in prev_dup:
            result_table[trows][tcols] = select_table[count]
            prev_dup.append(position)
        else:
            count-=1
        count += 1

store_combi = list(chain.from_iterable(result_table))
for i, v in enumerate(result_table):
    rt = 0
    while rt < len(v):
        ran_key_names = key_name[random.randrange(0, len(key_name))]
        if v[rt] == None:
            if store_combi.count(ran_key_names) < 2:
                result_table[i][rt] = ran_key_names
                store_combi.append(ran_key_names)
            else:
                rt-=1
        rt+=1 
pprint(result_table)